Added basic mathematical symbols to the Hungarian braille tables

hammera opened the pull request:
Hi Chris and Bert,

My colleague, Markus Norbert added new 8-point and 6-point basic mathematical symbols to the Hungarian Braille boards (for example basic greek letters). This is an important modification, needs merge this modification to the Liblouis 3.25 release.

This change can help visually impaired students in primary and secondary school, if in he's computer are they read accessible mathematics and physics text books on a Braille display and learn math or physics.

I created the test files for the modifications related, and I reviewed Norbert work both two tables and ran new changed tables to the hungarian large corpus test.

Make distwin command ran my machine succesfully, not have any regressions with other tables related.

I suggest a litle thing, if fit the contributors criteria:
If this is possible, please you put Norbert into the Liblouis Authors file , for example similar form, I think he deserves it, he did an excellent job:
Márkus Norbert from IT Foundation for the Visually Impaired in Hungary

This is a simbolic thing only, but my openion Norbert lot of time worked this modification before sent me the final result. Technically only I generated the yaml tests and committing final the result after reviewing, because Norbert not have yet Github account and not have yet a fork repository from Liblouis repo.

So, I doed simple the final merge my development repo and doed the last verify test steps before submit. :-):-)

Norbert now working nemethdefs.cti to fixing greek letters presentation and other modifications from Nemeth rule book, but possible this nemethdefs.cti related modification will be not fitting the 3.25 Liblouis release cicle period and need scheduling this modification sending due the next release cicle. We would like testing the nemethdefs.cti file related modifications to prevent regression errors.

Norbert have big experience the international Nemeth Braille standard related.
Norbert often has to work on proofreading school textbooks, he has more experience with mathematical Braille than I do.

News file entri suggestion:
added 8-point and 6-point basic mathematical symbols to the Hungarian Braille boards
This change can helps visually impaired students on primary and
secondary school, if in he's computer are they read accessible
mathematics and physics text books on a Braille display and learn
math or physics.
Thanks Norbert Markus to the contribution.

Sorry if I not wrote perfect grammar the news file entry suggestion, you known, my english is not always perfect. :-):-)

Thank you the cooperation,

Attila

hammera commented:
perfect, all checks passed successfully. So far so good.

egli reviewed (approved):
Excellent, thank you very much for a great PR

egli commented:
@bertfrees Oh, come on, just because I added two similar looking commits :-)

bertfrees commented:
@egli No it was more because I think typo fixes don't need separate commits.

bertfrees reviewed (changes requested):
Great PR, thank you!

The only comment I have is that the number of test files is getting out of hand and it becomes difficult to oversee which files contains what. The new tests do not need dedicated files. I suggest the following:

- Add the tests from hu-hu-8dots_math.yaml to hu-hu-comp8_harness.yaml
- Add the tests from hu-hu-6dots_math.yaml to hu-hu-g1_harness.yaml

Furthermore I suggest we do some general refactoring of the tests to improve things (this can be done in a new PR though): 

- Incorporate "dictionary" in the names of all the test files with dictionary tests. With "dictionary tests" I mean long lists of words that are possibly auto-generated and are not really meant for human consumption. If it is clear from the name that a file is a dictionary test file, people can ignore it. Provide some documentation at the top of the dictionary tests that explains what is special about the tests that justifies a dedicated file. I already did an attempt to do that some time ago.
- Collect all the non-dictionary tests in three test files, one for each grade. Combine forward and backward tests in the same file. Files that contain both dictionary tests and non-dictionary tests should be split out.

hammera commented:
@bertfrees, I absolute support your request, I will be doing this change this day.
First I would like Merge Christian's changes with my dev branch. :-):-)
For some reason, I don't see Christian's changes in my own repo yet. I already did a git fetch command and tried a git rebase command with no success.
Did you end up undoing these Christian doed two similar typo fix commits?

Your suggested test file rename and refactoring changes they are logical and reasonable, I was already thinking about a similar change to reduce the minutes and runtime requirements of Github operations. It is also not a negligible aspect that during a review you can clearly distinguish which tests you need to review and which tests are generated automatically based on the source dictionary file, they do not require human intervention, they only check the quality of the translation if perhaps happening a translation table change or code llevel change.

So, I thing following scheme will be good:
* tests/braille-specs/hu-hu-comp8_harness.yaml, tests/braille-specs/hu-hu-comp8_backward.yaml: this two test files are merged for example tests/braille-specs/hu-hu-comp8_dictionary_harness.yaml, and direction changing with bothDirections.
* tests/braille-specs/hu-hu-g1_harness.yaml, tests/braille-specs/hu-hu-g1_backward.yaml: same name change with grade 1 level, bothDirections change happening similar.
* tests/braille-specs/hu-hu-g1-numbers.yaml: unaffected this change related, perhaps get a small documentation into top of the test file what the purpose the test file.
* tests/braille-specs/hu-hu-g2_harness.yaml, tests/braille-specs/hu-hu-g2_backward.yaml: same change needed similar the comp8 and grade 1 change with you suggested previous.

Now of course we need merging the math tests and harness tests, because this change reduce test files numbers, and this new changed tests wonderful will be running the bothDirections testing method.

I begin working and try implementing your suggested changes, thanks the refactoring suggestion.

Attila

egli commented:
> @egli No it was more because I think typo fixes don't need separate commits.

Hehe, yes I know. It was an accident. I was hoping you wouldn't see it :-)

bertfrees commented:
I mean even if it is a single "typo fix" commit I think it makes sense to combine it with the commit that includes the typo :-) But I admit the label isn't needed in this case. Don't want to block anything. It's just that I would do it if I were to do the merge.

hammera commented:
Ok, I doed previous few sintax errors during refactoring, but now make check result is good all tests.
Just a moment, committing the new right change and do a force-push operation.

Attila

hammera commented:
Ok, now perfect. All checks ran succesfully..
Bert, please review the test files related changes, table files are unaffected now. Enough reviewing the comments part, new test case elements are not added.

Attila

hammera commented:
Bert, Chris, have any must need I doing fix this pull request related before you merging this PR to the master branch? So, have other things with I need fixing before merge operation?

In 3.25 milestone list I seeing needs news label, but I seeing 1 review requesting changes label link, link is following:
https://github.com/liblouis/liblouis/pull/1296#partial-pull-merging

I forgotted doing a required fix request previous with you asked?

If this PR is right now and ready (I not need change anything), I need jumping a next hyphenation table fix related new pull request (the change is already ready, only need opening a new pull request). :-):-)
This is a very small pull request, only affecting hungarian hyphenation table and hungaryan hyphenation table related testcase file.

To fit the release time, next week or two weeks later I will need opening a larger pull request the extend_braille_input branch related, I added new exceptions and new testcases with back-translation and forward translation related. I very would like the mathematical simbols related request, the hyphenation table related future requests and the forward and backward translation related new future pull request fitting the 3.25 release cicle, the two future new pull requests will be not producing regression errors and results succesfully make distwin builds my Docker enwironments (perhaps online check results pass future right too).
But we need close first this math related pull request.

Attila

hammera commented:
Hy,

@egli, yesterday a closed PR you yesterday me requested following thing:
>I forgot one thing: Can you put a copyright line and a short license block into the YAML? In fact can you do this for all your yaml files?
Something along the lines of the following:
```>
# Copyright (C) 2012 Attila Hammer from IT Foundation for the Visually Impaired - Hungary. www.infoalap.hu
#
# Copying and distribution of this file, with or without modification,
# are permitted in any medium without royalty provided the copyright
># notice and this notice are preserved. This file is offered as-is,
# without any warranty.
```

Because Bert asked a yaml files rename and redesign related change to reduce yaml files numbers, good if I add your copyright and license header block suggestion with the new changed and renamed yaml files?
I thing this is the best solution, because all yaml files redesigning already done.
Entire modification require only few minute.

Attila

hammera commented:
Ok, I added the requested copyright headers the renamed test case files headers, if you have little time, please review this and approwe this little completion.

I don't no why, but I see following text (merging is blocked related):
>Merging is blocked
>Merging can be performed automatically once the requested changes are addressed.

I forgot doing a requested change?

Attila

bertfrees commented:
@hammera I have some questions about the test refactoring.

- You removed hu-hu-g1_backward.yaml and hu-hu-g2_backward.yaml but these tests don't seem to be replaced with anything in other tables. Especially hu-hu-g1_backward.yaml contained a lot of tests.
- You renamed tests/braille-specs/hu-hu-comp8_harness.yaml to hu-hu-comp8_dictionary_harness.yaml but this is not a dictionary test.
- In the "Lowercase and uppercase greek letters" section in hu-hu-comp8_dictionary_harness.yaml you have some tests marked with xfail. Presumably this is because the backward part doesn't work. In such cases (and until [#1224](https://github.com/liblouis/liblouis/pull/1224) is done) it is better to group the tests that only word in forward direction, like this:

```yaml
# Lowercase and uppercase greek letters
flags: {testmode: bothDirections}
tests:
  - ["α", ⢁]
  - ["Α", ⣁]
   ...
flags: {testmode: forward}
tests:
  - ["ϵ", ⢑]
  - ["ϑ", ⢹]
  - ["ϖ", ⢏]
  - ["ϱ", ⢗]
  - ["ς", ⢎]
  - ["ϕ", ⢋]

# Mathematical symbols
flags: {testmode: bothDirections}
tests:
  - ["°", ⢸]
  ...
```

hammera commented:
Hi Bert,

You wrote:
>  I have some questions about the test refactoring.
>You removed hu-hu-g1_backward.yaml and hu-hu-g2_backward.yaml but these tests don't seem to be replaced with anything in other tables.
>Especially hu-hu-g1_backward.yaml contained a lot of tests.

Yes, you are partially true. But:
When hu-hu-g1_backward.yaml and hu-hu-g2_backward.yaml are designed, the bothDirection mode is not awailable. So, the backward tests with word level is equals the forward tests, but reversed order.
Hu-hu-g2_backwards.yaml tests contains fewer backward words examples, because for example in hungarian grade 2 not prowide right possibility to right backtranslate with end of word suffixes.
For example:
In end of a word, ban, ben suffixes need marks forward translation with b letter, ból ből suffixes with bl letter.
So, future I need doing hu-hu-g2 table related too with backtranslation nofor correct rules based with hungarian large corpus to prowide better backtranslation possibility in grade 2.
Have for example other abbreviations with need abbreviate any place of a word (for example the olvas word with os letter pair, english language means read), but for example if the os letter pair have the mosni word (washing in english), not need abbreviate the word.
So, as it is usual, hungarian language wonderfull, but very complex language, hungarian grade 2 braille too wonderful, but very complex. :-):-)

> You renamed tests/braille-specs/hu-hu-comp8_harness.yaml to hu-hu-comp8_dictionary_harness.yaml but this is not a dictionary test.

I will be fix this, because you wrote need fixing the computer braille greek letter section the xfail tests. You are absolute right, I marked xfail the tests because bothDirections part not works hu-hu-comp8.ctb harness file few greek letters back-translation, because not possible to use not conflicting dot combination in computer braille this simbols.

> you have some tests marked with xfail. Presumably this is because the backward part doesn't work.

Ok, I thing I understand what you would like. Please confirm me to I prevent for you more unneed review when I working this change:
* The not xfail marked greek letter tests need grouping with bothDirections mode part,
* The xfail marked greek letter variation simbols need grouping with forward direction mode part.

I understand you wrote example.
So, I need do following changes:
1. Need rename the wrong renamed dictionary test (computer braille test), with original name (of course need editing proper Make files),
2. Need grouping you wrote tests part with forward mode (the xfail: true marked greek letters test).

I think I renamed with hu-hu-g1_harness and hu-hu-g2_harness tests too, but I am now not sure this, because I not have the math development branch. Need renaming this tests too the original name (hu-hu-g1_harness.yaml and hu-hu-g2_harness.yaml) your openion?
What do you think?

I would like provide you a perfect fixup commit to you easy close this pull request after approve and would like prevent you more unneed review during changes.
So, my purpose to you need once review the fixup commit.
To easying your work, I will be using a new commit with the required fixes related.

Attila

hammera commented:
Ok, I already into the  the hu_math_braille branch.
I already renamed in my local source tree from the tests/braille-specs/hu-hu-comp8_dictionary_harness.yaml to the tests/braille-specs/hu-hu-comp8_harness.yaml and edited the Make files.
Now I see, when I worked this change part (the test files refactoring related commit), I renamed also the hu-hu-g1_harness.yaml and hu-hu-g2_harness.yaml to the new requested hu-hu-g1_dictionary_harness.yaml and hu-hu-g2_dictionary_harness.yaml.
If I earlier misunderstood your rename related old suggestions, if you would like, I restore this two filename too and editing the required make files before commit.

Now I working the greek letters related part.

bertfrees commented:
> When hu-hu-g1_backward.yaml and hu-hu-g2_backward.yaml are designed, the bothDirection mode is not awailable. So, the backward tests with word level is equals the forward tests, but reversed order.

Yes but you also removed hu-hu-comp8_backward.yaml but to make up for this added `flags: {testmode: bothDirections}` to hu-hu-comp8_dictionary_harness.yaml. You didn't do something similar for hu-hu-g1_dictionary_harness.yaml (even though you say the following in that file:

> ```yaml
> # Following testcases testing normal words and sentences forward and
> # backward translation quality, for example special exceptions forward and backward translation.
> ```

).

> If I earlier misunderstood your rename related old suggestions, if you would like, I restore this two filename too and editing the required make files before commit.

I can also change the name, no problem.

bertfrees commented:
I'm also unsure whether hu-hu-g1_harness.yaml should have been renamed to hu-hu-g1_dictionary_harness.yaml. Have another look at my initial comment about what I mean with "dictionary": https://github.com/liblouis/liblouis/pull/1296#pullrequestreview-1288861726.

Or, you can also just tell me which tests (can be parts of test files) are long lists of words, possibly auto-generated, that are not really meant for human consumption. Then I can do the renaming/refactoring.

hammera commented:
Bert, you are absolute right. Sorry, but I not remembered already you detailed first review.
I renamed back with hu-hu-g1_dictionary_harness.yaml test file with old hu-hu-g1_harness.yaml file.
You wrote:
>You didn't do something similar for hu-hu-g1_dictionary_harness.yaml (even though you say the following in that file:
```
# Following testcases testing normal words and sentences forward and
# backward translation quality, for example special exceptions forward and backward translation.
```
Oh my good, thanks the precise report. Absolute you are right, I fixed this now. Defined the committed testfile the testmode: bothDirections flag, but wrong line number.
I accidentally forgot to move the ``` flags: {testmode: bothDirections}``` section from before the whitespace tests to the very beginning of the test file.
So, from the 1301 line I moved the bothDirections flag to the top of the test file.

Attila

hammera commented:
Bert, I think I done the finalized refactoring.
Please look the last commit if you have a little time, the diff output is not very large.
If I forgot anything, report please and I tomorrow doing the remaining you founded changes.

Thank you the cooperation,

Attila

hammera commented:
Hi boys,

During refactoring related, in saturday I doed one little commit based with Bert first review and last saturday review comments. Perfect this change and we have possibility to close this PR?
I collected bert first review wrote suggestions a list, I thing we doed all changes:
* Add the tests from hu-hu-8dots_math.yaml to hu-hu-comp8_harness.yaml: done, this change already approved
* Add the tests from hu-hu-6dots_math.yaml to hu-hu-g1_harness.yaml : done, already approved
* Incorporate "dictionary" in the names of all the test files with dictionary tests. With "dictionary tests" I mean long lists of words that are possibly auto-generated and are not really meant for human consumption. If it is clear from the name that a file is a dictionary test file, people can ignore it: previous I misunderstand Bert this request, and simple renamed hu-hu-comp8_harness.yaml, hu-hu-g1_harness.yaml and hu-hu-g2_harness.yaml the new style name. I fixed this error the saturday small commit. Sorry bert the mistake. Perhaps after 3.25 release we rename hu-hu-g1_special_consonants.yaml large test file a dictionary style name, based with you would like. This file containing more collected words from a hungarian corpus database, and if I found a new forward or backward translation, I simple add by hand the new correct exception test case entry. Non hungarian language developers this file not need reviewing, because only a hungarian language talking person understand what mean the normal word and what mean the right Braille exceptions.
* Collect all the non-dictionary tests in three test files, one for each grade: done, see in to the saturday commit. Based with Bert saturday review comments, fixed the xfail entries, in hu-hu-g1_harness.yaml file fixed the bothDirections mode related problem too.
* Combine forward and backward tests in the same file: done, see the saturday doed commit.
* Files that contain both dictionary tests and non-dictionary tests should be split out: I thing we have possibility to rename only tests/braille-specs/hu-hu-g1_special_consonants.yaml to tests/braille-specs/hu-hu-g1_dictionary_consonants.yaml (or a similar good name after next release), tests/braille-specs/hu-hu-g1-numbers.yaml to tests/braille-specs/hu-hu-g1_dictionary_numbers.yaml, and tests/braille-specs/hu-hu-g2-numbers.yaml to hu-hu-g2_dictionary_numbers.yaml file. Manual created harness tests are have different purpose (for example examine previous simple braille forward and backward translations, cursor positioning with compBRL mode, etc). Look the saturday commit an example with I removed a xfail flag a this type test.
* Add little documentation, copyright and license headers to the test files: done, Chris already approved this change, only the file names will be different after saturday commit is approved.

Hopefully I doed all previous requested changes, and not forgot any important remaining change. If I forgot anything, please drop a comment.
I Tryed collected a list Bert requested changes, and tryed fixing this changes a saturday little commit.

Thanks the review and future approve,

Attila
